src/inmanta/server/services/orchestrationservice.py

```diff
@@ -1032,87 +1032,96 @@ async def release_version(
-            await model.mark_done(connection=connection)
-            return 200, {"model": model}
+                increments: tuple[abc.Set[ResourceIdStr], abc.Set[ResourceIdStr]] = await self.resource_service.get_increment(
+                    env,
+                    version_id,
+                    connection=connection,
+                )
 
-        if push:
-            # fetch all resource in this cm and create a list of distinct agents
-            agents = await data.ConfigurationModel.get_agents(env.id, version_id, connection=connection)
-            await self.autostarted_agent_manager._ensure_agents(env, agents, connection=connection)
-
-            for agent in agents:
-                client = self.agentmanager_service.get_agent_client(env.id, agent)
-                if client is not None:
-                    if not agent_trigger_method:
-                        env_agent_trigger_method = await env.get(ENVIRONMENT_AGENT_TRIGGER_METHOD, connection=connection)
-                        incremental_deploy = env_agent_trigger_method == const.AgentTriggerMethod.push_incremental_deploy
+                increment_ids, neg_increment = increments
+                await self.resource_service.mark_deployed(env, neg_increment, now, version_id, connection=connection)
+
+                # Set the updated field:
+                await data.Resource.copy_last_success(env.id, version_id, connection=connection)
+
+                # Setting the model's released field to True is the trigger for the agents to start pulling in the resources.
+                # This has to be done after the resources outside of the increment have been marked as deployed.
+                await model.update_fields(released=True, result=const.VersionState.deploying, connection=connection)
+
+            if model.total == 0:
+                await model.mark_done(connection=connection)
+                return 200, {"model": model}
+
+            if push:
+                # fetch all resource in this cm and create a list of distinct agents
+                agents = await data.ConfigurationModel.get_agents(env.id, version_id, connection=connection)
+                await self.autostarted_agent_manager._ensure_agents(env, agents, connection=connection)
+
+                for agent in agents:
+                    client = self.agentmanager_service.get_agent_client(env.id, agent)
+                    if client is not None:
+                        if not agent_trigger_method:
+                            env_agent_trigger_method = await env.get(ENVIRONMENT_AGENT_TRIGGER_METHOD, connection=connection)
+                            incremental_deploy = env_agent_trigger_method == const.AgentTriggerMethod.push_incremental_deploy
+                        else:
+                            incremental_deploy = agent_trigger_method is const.AgentTriggerMethod.push_incremental_deploy
+                        self.add_background_task(client.trigger(env.id, agent, incremental_deploy))
```

Comment:
I think we may need to first commit the transaction before we add any background tasks, e.g. dedent the whole `if push` block to live outside of the transaction.

Reply:
it is outside the transaction (but on the same connection)

Reply:
Oh good, sometimes difficult to see.

One potential problem I still see (though it existed before this PR as well I think) is that if the caller passes in a collection object, wrapped in a transaction, we add these background tasks before that outer connection commiits.

Reply:
Yes, I think that is part of the responsibility of the caller.

Passing along connections with open transactions is not something we can consider to be modular.

We could assert there is no ongoing transaction, but that is perhaps a bit harsh?

Reply:
added it